### PR TITLE
feat: use vim.treesitter.language.add()

### PR DESCRIPTION
**Pull request by lewis6991**
Requires https://github.com/neovim/neovim/pull/22207

### CODE REVIEW

**Comment by theHamsta** [comment]
Setting to draft because we need this is Neovim release to be able to merge.

What is the motivation of mixing ft and lang? Tree-sitter languages are a different concept from Neovim filetype. The user of `require_lang` should know how which `.so` they want. We have a Lua table that maps ft to lang where they don't coincide. Your upstream PR would essentially move the role of this Lua table to the C hashtable that Neovim `treesitter.c` has. Note that this will cause loading the parser multiple times when you're requiring it from different languages.

If you don't want to use nvim-treesitters highlight module you can have you own Lua table that decides which treesitter lang should apply to which filetype. Neovim could also have such a ft->lang Lua table upstreamed. But I wouldn't make `require_language` try out filenames. The user of this function should know what they want. If the filename of the `.so` is an additional concept that needs to be configured it could be an additional parameter. But I think that having the filename of the `.so` as a key in the `treesitter.c` language hashmap is good as both the SO-file in rtp and the key in the hashmap are unique then.

**Comment by lewis6991** [comment]
I've iterated through the upstream PR quite a bit. Ended up adding `vim.treesitter.language.add()` which has the ability to associate filetypes to langs.

**Review by theHamsta — APPROVED** [review]
Thank you! I really like how much cleaner the API in Neovim is compared to the previous approach we had here!

File: README.md [hunk 1]
```diff
@@ -478,11 +478,10 @@ parser_config.zimbu = {
 }
 ```
 
-If you wish to set a specific parser for a filetype, you should extend the `filetype_to_parsername` table:
+If you wish to set a specific parser for a filetype, you should use `vim.treesitter.language.register()`:
```
Comment (theHamsta) [inline README.md:481]: @lewis6991 this should maybe be documented that this is still nightly

File: lua/nvim-treesitter/parsers.lua [hunk 2]
```diff
@@ -45,20 +71,12 @@ local list = setmetatable({}, {
...
-          if key == "used_by" then
-            require("nvim-treesitter.utils").notify(
-              "used_by is deprecated, please use 'filetype_to_parsername'",
-              vim.log.levels.WARN
-            )
-            filetype_to_parsername[value] = parsername
-          else
-            rawset(parserconfigtable, key, value)
-          end
+          rawset(parserconfigtable, key, value)
```
Comment (theHamsta) [inline lua/nvim-treesitter/parsers.lua:74]: Is the list a normal table now? I think the code can  be removed then

**Comment by clason** [comment]
I agree with @theHamsta that although this is a very welcome and necessary change, we should either gate it behind a version check or wait until 0.9 is released and do a proper deprecation/warning cycle.
(Especially since there are a number of upstream changes that should be reflected in nvim-treesitter, like fold upstreaming and `get_node` consolidation.)

Would it make sense to open a new branch targeting master and doing all the refactors there until the release?

**Comment by lewis6991** [comment]
Note the PR explicitly checks for `vim.treesitter.language.register` and will continue to work with v0.8., which is why CI passes. I've only removed documentation about the `filetype_to_parsername` table, but it still exists.

**Comment by clason** [comment]
Should we remove the _old_ deprecation warning while we're at it, then?

File: lua/nvim-treesitter/parsers.lua [hunk 3]
```diff
@@ -1,7 +1,31 @@
...
+
+local filetype_to_parsername = {}
+
+if new_lang_api then
+  filetype_to_parsername = setmetatable({}, {
+    __newindex = function(_, k, v)
+      require("nvim-treesitter.utils").notify(
+        "filetype_to_parsername is deprecated, please use 'vim.treesitter.language.register'",
+        vim.log.levels.WARN
+      )
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:14]: this warning ☝️

Comment (lewis6991) [reply lua/nvim-treesitter/parsers.lua:14]: This warning only appears if `vim.treesitter.language.register` is present, meaning you are running vim v0.9.

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:14]: No, I mean it's been around long enough, and using `filetype_to_parsername` can be turned into a hard error. (No need to carry around deprecations forever.)

Comment (lewis6991) [reply lua/nvim-treesitter/parsers.lua:14]: I thought `filetype_to_parsername` was the recommended way for nvim v0.8 which replaces `used_by` in the parser-config, which warning was removed in this PR.

Comment (lewis6991) [reply lua/nvim-treesitter/parsers.lua:14]: This has been removed:
```lua
        __newindex = function(parserconfigtable, key, value)
          if key == "used_by" then
            require("nvim-treesitter.utils").notify(
              "used_by is deprecated, please use 'filetype_to_parsername'",
              vim.log.levels.WARN
            )
            filetype_to_parsername[value] = parsername
          else
            rawset(parserconfigtable, key, value)
          end
        end,

```

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:14]: Yes, that was a clear reading fail on my end; I didn't notice the changed message, just saw the deprecation as being moved. 🤦

**Comment by lewis6991** [comment]
Done.

**Comment by clason** [comment]
Before you open an issue, it would be helpful to try to replicate this without nvim-treesitter (just put parser and queries into your own config directory).

**Pull request merged**